### PR TITLE
Sync database to coda

### PR DESCRIPTION
**Pull request by as2388**
Review #4 first.

Adds minimum viable implementation of a sync from the database to Coda. I haven't added the reverse (coda -> database), but that should be straightforward (when a message changes in Coda, get the messages from the database with the same dataset and id, and apply the same _update_db_message_from_coda_message routine included here.

### CODE REVIEW

**Review by IsaackMwenda — APPROVED** [review]

File: src/engagement_db_to_coda/engagement_db_to_coda.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+    :param transaction: Transaction in the engagement database to perform the update in.
+    :type transaction: google.cloud.firestore.Transaction
+    :param engagement_db: Engagement database to sync from.
+    :type engagement_db: engagement_database.EngagementDatabase
+    :param coda: Coda instance to sync the message to.
+    :type coda: coda_v2_python_client.firebase_client_wrapper.CodaV2Client
+    :param message_id: Id of the message to sync.
+    :type message_id: str
+    """
+    db_message = engagement_db.get_message(message_id, transaction=transaction)
```
Comment (DanielMwendwa) [inline src/engagement_db_to_coda/engagement_db_to_coda.py:141]: I see db refers engagement database but consider renaming db_message to engagement_db_message /  engmt_db_msg

File: src/engagement_db_to_coda/engagement_db_to_coda.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+    :type coda_config:  src.engagement_db_to_coda.configuration.CodaSyncConfiguration
+    :param transaction: Transaction in the engagement database to perform the update in.
+    :type transaction: google.cloud.firestore.Transaction | None
+    """
+    coda_dataset_config = coda_config.get_dataset_config_by_engagement_db_dataset(db_message.dataset)
+
+    # Check if the labels in the engagement database message already match those from the coda message.
+    # If they do, return without updating anything.
+    # TODO: Validate if the dataset is correct too?
+    if json.dumps([x.to_dict() for x in db_message.labels]) == json.dumps([y.to_dict() for y in coda_message.labels]):
```
Comment (DanielMwendwa) [inline src/engagement_db_to_coda/engagement_db_to_coda.py:79]: Can we add dunder eq to core_data_modules.data_models.Label?

File: src/engagement_db_to_coda/engagement_db_to_coda.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+    :param coda: Coda instance to add the message to.
+    :type coda: coda_v2_python_client.firebase_client_wrapper.CodaV2Client
+    :param coda_dataset_config: Configuration for adding the message.
+    :type coda_dataset_config: src.engagement_db_to_coda.configuration.CodaDatasetConfiguration
+    :param db_message: Message to add to Coda.
+    :type db_message: engagement_database.data_models.Message
+    """
+    log.debug("Adding message to Coda")
+
+    coda_message = core_data_modules.data_models.Message(
```
Comment (DanielMwendwa) [inline src/engagement_db_to_coda/engagement_db_to_coda.py:30]: import Message at the top to avoid the repetition of core_data_modules.data_models

**Review by DanielMwendwa — APPROVED** [review]
Thanks, @as2388.

File: configurations/test_pipeline_configuration.py [hunk 4]
```diff
@@ -1,7 +1,21 @@
-from src.pipeline_configuration_spec import PipelineConfiguration, RapidProSource
-from src.common.configuration import EngagementDatabaseClientConfiguration, UUIDTableClientConfiguration, RapidProClientConfiguration
+import json
+
+from core_data_modules.cleaners import swahili
+from core_data_modules.data_models import CodeScheme
+
+from src.common.configuration import RapidProClientConfiguration, CodaClientConfiguration, UUIDTableClientConfiguration, \
+    EngagementDatabaseClientConfiguration
+from src.engagament_db_to_coda.configuration import CodaSyncConfiguration, CodaDatasetConfiguration, \
```
[the remark this answers is not in the document]
Comment (as2388) [reply configurations/test_pipeline_configuration.py:8]: Thanks for catching this. It looks like a merge error, I've corrected and tested now

File: src/engagement_db_to_coda/engagement_db_to_coda.py (hunk 3 above)
Comment (as2388) [reply src/engagement_db_to_coda/engagement_db_to_coda.py:30]: I think this needs to be explicit, to make super clear that this is a Coda message and not an engagement db message - I'll add a comment in the code

Comment (as2388) [reply src/engagement_db_to_coda/engagement_db_to_coda.py:30]: Actually I've changed this to `from core_data_modules.data_models import Message as CodaMessage` at the top, so we're not doing any dynamic imports

File: src/engagement_db_to_coda/engagement_db_to_coda.py (hunk 2 above)
Comment (as2388) [reply src/engagement_db_to_coda/engagement_db_to_coda.py:79]: Added here https://github.com/AfricasVoices/CoreDataModules/pull/220, will update here once reviewed

File: src/engagement_db_to_coda/engagement_db_to_coda.py (hunk 1 above)
Comment (as2388) [reply src/engagement_db_to_coda/engagement_db_to_coda.py:141]: renamed to engagement_db_message, thanks

**Pull request merged**